src/main/scala/is/hail/methods/LinearMixedRegression.scala

```diff
@@ -184,97 +181,108 @@ object LinearMixedRegression {
       vds2
   }
 
-  def computeNEigs(S: DenseVector[Double], nEigs: Int, optDroppedVarianceFraction: Option[Double]): Int = {
-    val trace = S.toArray.sum
-    var i = S.length - 1
+  def computeNEigsDVF(S: DenseVector[Double], droppedVarianceFraction: Double): Int = {
+    require(0 <= droppedVarianceFraction && droppedVarianceFraction < 1)
+    val trace = sum(S)
+
+    var i = 0
     var runningSum = 0.0
-    if (optDroppedVarianceFraction.isEmpty) return nEigs
-    val target = optDroppedVarianceFraction.get * trace
-    while (i > nEigs || runningSum <= target) {
-      runningSum += S(-i)
-      i -= 1
+    val target = droppedVarianceFraction * trace
+    while (runningSum <= target) {
+      //Note that S is sorted smallest to largest
+      runningSum += S(i)
+      i += 1
     }
-    math.min(nEigs, i + 1)
+    //i is the number of eigenvalues it took to pass the target value. One too big.
+    i = i - 1
+    S.length - i
   }
 }
 
-trait PerVariantLMM {
+trait ScalerLMM {
   def likelihoodRatioTest(v: Vector[Double]): Annotation
 }
 
-//Works in any situation, but is slower than ScalerLMM
-class StandardLMM(lmmConstants: LMMConstants, delta: Double, logNullS2: Double, useML: Boolean) extends PerVariantLMM {
+// Handles full-rank case
+class FullRankScalerLMM(
+  y: DenseVector[Double],
+  yy: Double,
+  Qt: DenseMatrix[Double],
+  Qty: DenseVector[Double],
+  yQty: Double,
+  T: DenseMatrix[Double],
+  logNullS2: Double,
+  useML: Boolean) extends ScalerLMM {
+
+  def likelihoodRatioTest(x0: Vector[Double]): Annotation = {
+    val x = T * x0
+    val n = y.length
+    val Qtx = Qt * x
+    val xQtx: Double = (x dot x) - (Qtx dot Qtx)
+    val xQty: Double = (x dot y) - (Qtx dot Qty)
+
+    val b: Double = xQty / xQtx
+    val s2 = (yQty - xQty * b) / (if (useML) n else n - Qt.rows)
```

Comment:
move `val invDof = 1 / (if (useML) n else n - Qt.rows)` outside the test. this stands for inverse degrees of freedom.